source/core/owt_base/internal/TransportServer.cpp

```diff
@@ -0,0 +1,305 @@
+                this, boost::asio::placeholders::error));
+    }
+}
+
+void TransportServer::acceptHandler(const boost::system::error_code& ec)
+{
+    if (m_isClosed) {
+        return;
+    }
+    if (!ec) {
+        if (m_isSecure) {
+            ELOG_DEBUG("Start handshake %d");
+            m_sslSocket->lowest_layer().set_option(tcp::no_delay(true));
+            m_sslSocket->async_handshake(
+                boost::asio::ssl::stream_base::server,
+                boost::bind(&TransportServer::handshakeHandler, this, m_sslSocket,
+                    boost::asio::placeholders::error));
+            m_sslSocket.reset();
+        } else {
+            m_socket.set_option(tcp::no_delay(true));
+            int sessionId = m_nextSessionId++;
+            m_sessions[sessionId] =
+                std::make_shared<TransportSession>(
+                    sessionId, m_service, std::move(m_socket), this);
+            ELOG_DEBUG("Accept session %d", sessionId);
+            if (m_listener) {
+                m_listener->onSessionAdded(sessionId);
+            }
+        }
+        doAccept();
+    } else if (ec.value() != boost::system::errc::operation_canceled) {
+        ELOG_WARN("Accept error: %s", ec.message().c_str());
+        doAccept();
+    }
+}
+
+void TransportServer::handshakeHandler(std::shared_ptr<SSLSocket> sock,
+                                       const boost::system::error_code& ec)
+{
+    if (!ec) {
+        int sessionId = m_nextSessionId++;
+        m_sessions[sessionId] =
+            std::make_shared<TransportSession>(
+                sessionId, m_service, sock, this);
+        ELOG_DEBUG("accept secure session %d", sessionId);
+        if (m_listener) {
+            m_listener->onSessionAdded(sessionId);
+        }
+    } else {
+        ELOG_WARN("Error during handshake: %s", ec.message().c_str());
+    }
+}
+
+
+void TransportServer::listenTo(uint32_t port)
+{
+    if (m_acceptor.is_open()) {
+        return;
+    }
+    boost::system::error_code ec;
+    tcp::endpoint endpoint(tcp::v4(), port);
+    m_acceptor.open(endpoint.protocol());
+    m_acceptor.bind(endpoint, ec);
+    if (!ec) {
+        m_acceptor.listen(boost::asio::socket_base::max_connections, ec);
+        if (!ec) {
+            doAccept();
+        }
+    } else {
+        ELOG_WARN("TransportServer listenTo failed: %s", ec.message().c_str());
+    }
+}
+
+void TransportServer::listenTo(uint32_t minPort, uint32_t maxPort)
+{
+    if (m_acceptor.is_open()) {
+        return;
+    }
+    boost::system::error_code ec;
+
+    // find port in range
+    uint32_t portRange = maxPort - minPort + 1;
+    uint32_t port = rand() % portRange + minPort;
+
+    for (uint32_t i = 0; i < portRange; i++) {
+        if (m_acceptor.is_open()) {
+            m_acceptor.close();
+        }
+        m_acceptor.open(tcp::v4());
+        m_acceptor.bind(tcp::endpoint(tcp::v4(), port), ec);
+
+        if (ec != boost::system::errc::address_in_use) {
```

Comment:
maybe ec != 0 is better if you're not sure this is the only error that would be thrown.